File: leetcode-go/Q051-Q100/Q067_Add_Binary/Q067_Add_Binary.go
package Q067_Add_Binary

func addBinary(a string, b string) string {
	ans := []byte{}
	aa, bb := len(a)-1, len(b)-1
	carry := 0
	for aa >= 0 || bb >= 0 {
		sum := carry
		if aa >= 0 {
			sum += int(a[aa] - '0')
			aa--
		}
		if bb >= 0 {
			sum += int(b[bb] - '0')
			bb--
		}
		ans = append([]byte{byte('0' + sum%2)}, ans...)
		carry = sum / 2
	}
	if carry != 0 {
		ans = append([]byte{'1'}, ans...)
	}
	return string(ans)
}

File: leetcode-java/src/facebook/Q067_Add_Binary.java
package facebook;

public class Q067_Add_Binary {

    public String addBinary(String a, String b) {
        StringBuilder sb = new StringBuilder();
        int i = a.length()-1, j = b.length()-1;
        int carry = 0;
        while (i >= 0 || j >= 0 || carry > 0) {
            int sum = carry;
            if (i >= 0) {
                sum += a.charAt(i)-'0';
                i--;
            }
            if (j >= 0) {
                sum += b.charAt(j)-'0';
                j--;
            }
            sb.insert(0, sum%2);
            carry = sum/2;
        }
        return sb.toString();
    }

}
